### ISSUE TITLE
Better configuration handling

### ISSUE THREAD

**Issue by sstok** [gushphp/gush#176]
Daniel Gomes did a good job at making the Config class, but for the future (and introduction of new features) we'd properly want a Configuration system that also supports proper validation of both the user input and current configuration (config schema changes for example), the Symfony2 Config component is a great start for inspiration, using the Config Component is also an option but might by a little complex for interactive usage.

To greatest way would allow to use this with Command configuration (as described in #175). It should be possible to Build (automatically configure) an interactive configuration so plug-ins (like Versioneye) can require/allow configuration from the user.

In summary.

* Allow defining an configuration schema (like Symfony2 Config component) (should use caching)
   * Validation of both user-input, and current configuration.
   * Configure normalizer/transformer per option, 
* Allow easy building of an interactive Configuration command with the Configuration schema.

Note. This would require a compat layer for existing adapters, as they currently use `doConfiguration()`, adapters can then be slowly upgraded with 'ConfigurableAdapter Interface' (or something like that).
